src/lib.rs

```diff
@@ -11,6 +11,678 @@
+    authority_name: String,
+    // Indicates where client instance is running.
+    client_type: UPClientMqttType,
+}
+
+pub enum UPClientMqttType {
+    Device,
+    Cloud,
+}
+
+#[allow(dead_code)]
+impl UPClientMqtt {
+    /// Create a new UPClientMqtt.
+    ///
+    /// # Arguments
+    /// * `config` - Configuration for the mqtt client.
+    /// * `client_id` - Client id for the mqtt client.
+    pub async fn new(
+        config: MqttConfig,
+        client_id: UUID,
+        authority_name: String,
+        client_type: UPClientMqttType,
+    ) -> Result<Self, UStatus> {
+        let topic_listener_map = Arc::new(RwLock::new(HashMap::new()));
+
+        let ptr = topic_listener_map.clone();
+
+        let mqtt_protocol = if config.ssl_options.is_some() {
+            "mqtts"
+        } else {
+            "mqtt"
+        };
+
+        let mqtt_uri = format!(
+            "{}://{}:{}",
+            mqtt_protocol, config.mqtt_hostname, config.mqtt_port
+        );
+
+        let mqtt_cli = mqtt::CreateOptionsBuilder::new()
+            .server_uri(mqtt_uri)
+            .client_id(client_id)
+            .max_buffered_messages(100)
+            .create_client()
+            .map_err(|e| {
+                UStatus::fail_with_code(
+                    UCode::INTERNAL,
+                    format!("Unable to create mqtt client: {e:?}"),
+                )
+            })?;
+
+        mqtt_cli.set_message_callback(
+            move |cli: &AsyncClient, message: std::option::Option<paho_mqtt::Message>| {
+                UPClientMqtt::on_receive(cli, message, ptr.clone())
+            },
+        );
+
+        // TODO: Integrate ssl options when connecting, may need a username, etc.
+        let conn_opts = mqtt::ConnectOptionsBuilder::with_mqtt_version(MQTT_VERSION_5)
+            .clean_start(false)
+            .properties(mqtt::properties![mqtt::PropertyCode::SessionExpiryInterval => 3600])
+            .finalize();
+
+        mqtt_cli.connect(conn_opts).await.map_err(|e| {
+            UStatus::fail_with_code(
+                UCode::INTERNAL,
+                format!("Unable to connect to mqtt broker: {e:?}"),
+            )
+        })?;
+
+        Ok(Self {
+            mqtt_client: Arc::new(mqtt_cli),
+            topic_listener_map,
+            authority_name,
+            client_type,
+        })
+    }
+
+    /// Helper function that handles MQTT messages on reception.
+    ///
+    /// # Arguments
+    /// * `cli` - MQTT client that received the message.
+    /// * `message` - MQTT message received.
+    fn on_receive(
+        _cli: &AsyncClient,
+        message: Option<paho_mqtt::Message>,
+        topic_map: Arc<RwLock<HashMap<String, HashSet<ComparableListener>>>>,
+    ) {
+        if let Some(msg) = message {
+            let topic = msg.topic();
+
+            // Get attributes from mqtt header.
+            let uattributes =
+                UPClientMqtt::get_uattributes_from_mqtt_properties(msg.properties()).unwrap();
```

Comment:
Would this cause a run-time panic if we got back the err path of the Result?

Reply:
Updated this to throw out a warning if its unable to parse the uattributes and return rather than just unwrapping.